### PR TITLE
MatrixFree: add DoFInfo::dof_indices_contiguous_sm

### PR DESCRIPTION
**Pull request by peterrum**
... and introduce a new communicator in `MatrixFree::TaskInfo` for SM.

~Depends on #11105.~

### CODE REVIEW

**Comment by peterrum** [comment]
@kronbichler I have pushed the first draft (not tested yet). Maybe you could take a quick look at the code and judge if what I am doing is not too naive and there are no obviously wrong assumptions.

File: include/deal.II/matrix_free/dof_info.h
Comment (peterrum) [inline include/deal.II/matrix_free/dof_info.h:248]: What would be a better name?

File: source/base/mpi_consensus_algorithms.cc
Comment (peterrum) [inline source/base/mpi_consensus_algorithms.cc:59]: With this we don't need the include in the other files anymore ;)

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (peterrum) [inline include/deal.II/matrix_free/matrix_free.templates.h:1575]: I guess it would be more safe to use `CellAccessor::global_active_cell_index()` and `CellAccessor::global_level_cell_index()` from #10490.

File: include/deal.II/matrix_free/dof_info.h
Comment (kronbichler) [reply include/deal.II/matrix_free/dof_info.h:249]: How about `identify_shared_memory_ranges` or `_regions`?

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h:1481]:
```suggestion
    // non-buffering mode is only supported if the indices of all cells are
    // contiguous for all dof_info objects.
```

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h:1498]: Do you intend to work on this TODO here or elsewhere?

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h:1640]: Is the commented-out part still needed?

**Review by kronbichler — APPROVED** [review]
Looks good to me with some minor adjustments. Do you still plan for additions, given that you have the "Do not merge" label still on this?

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (peterrum) [reply include/deal.II/matrix_free/matrix_free.templates.h:1498]: I would like to move this - in a follow-up PR - to `Utilities::MPI`. Does this make sense or is this function too specialized? I am using practically these lines also in #11098.

**Comment by peterrum** [comment]
@kronbichler I have updated the code (and remove the `Do not merge` label).

File: include/deal.II/matrix_free/matrix_free.templates.h
Comment (kronbichler) [reply include/deal.II/matrix_free/matrix_free.templates.h:1498]: I think such a function would be of general use, so the `MPI` namespace seems right.

File: include/deal.II/matrix_free/dof_info.h
Comment (kronbichler) [inline include/deal.II/matrix_free/dof_info.h:251]:
```suggestion
      compute_shared_memory_contiguous_indices(
```

**Comment by kronbichler** [comment]
I will merge once you update the typo in the function name.

**Comment by peterrum** [comment]
@kronbichler fixed!

**Pull request merged**
